### PR TITLE
Media type plugins

### PR DESCRIPTION
**Pull request by slashrsm**
Initial implementation. Should be ready for review. Field map UI is a follow-up (https://www.drupal.org/node/2336157).

https://www.drupal.org/node/2335645

Also done as part of this (test implementation): https://github.com/drupal-media/media_entity_youtube

### CODE REVIEW

File: src/MediaBundleForm.php [hunk 1]
```diff
@@ -51,21 +53,46 @@ public function form(array $form, FormStateInterface $form_state) {
...
+      '#description' => t('Media type provider plugin that is responsible for additional logic related to this media.'),
+    );
+
+    $form['type_configuration'] = array(
+      '#type' => 'fieldset',
+      '#title' => t('Type provider configuration'),
+      '#tree' => TRUE,
+    );
+
+    foreach ($plugins as $plugin => $definition) {
```
Comment (marcingy) [inline src/MediaBundleForm.php:83]: We have too foreach loops can they be merged in some way? Not a biggie but seems 'wrong'.

File: src/MediaTypeException.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+   *   [optional] The Exception message to throw.
+   * @param int $code
+   *   [optional] The Exception code.
+   * @param \Exception $previous
+   *   [optional] The previous exception used for the exception chaining.
+   */
+  public function __construct($element = NULL, $message = "", $code = 0, \Exception $previous = null) {
+    parent::__construct($message, $code, $previous);
+    $this->element = $element;
+
```
Comment (marcingy) [inline src/MediaTypeException.php]: Extra blank line

File: src/MediaTypeInterface.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Returns the display label.
+   *
+   * @return string
+   *   The display label.
+   */
+  public function label();
+
+  public function providedFields();
```
Comment (marcingy) [inline src/MediaTypeInterface.php:31]: Missing doxygen

File: src/MediaTypeInterface.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Get's a media-related field/value.
+   *
+   * @param MediaInterface $media
+   *   Media object.
+   * @param $name
+   *   Name of field to fetch.
+   * @return mixed
```
Comment (marcingy) [inline src/MediaTypeInterface.php:41]: Missing new line

File: src/MediaTypeInterface.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+   *   Field value or FALSE if data unavailable.
+   */
+  public function getField(MediaInterface $media, $name);
+
+  /**
+   * Gets settings (sub) form for type plugin.
+   *
+   * @param MediaBundleInterface $bundle
+   *   Media bundle.
+   * @return mixed
```
Comment (marcingy) [inline src/MediaTypeInterface.php:52]: Missing new line

File: src/MediaBundleForm.php (hunk 1 above)
Comment (slashrsm) [inline src/MediaBundleForm.php:83]: Hm... would be nice but merging those two loops would probably result in some strange code (defining #options after form element definition, ....). I am afraid doing that would actually result in less readable code. 

In terms of performance it shouldn't have any significant impact as it is O(n) in both cases. :)

**Pull request merged**